### PR TITLE
ZCS-3515: Implement Facebook oauth2 handler

### PR DESCRIPTION
**Pull request by jdavid-synacor**

### CODE REVIEW

File: src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+    // store refreshToken
+    oauthInfo.setUsername(username);
+    oauthInfo.setRefreshToken(credentials.get("access_token").asText());
+    dataSource.syncDatasource(mailbox, oauthInfo);
+    return true;
+  }
+
+  /**
+   * Validates that the response from authenticate has no errors, and contains the
```
Comment (desouzas) [inline src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java]: Delete this orphan docblock.

File: src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+      request.setQueryString(queryString);
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the authorization token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring an authorization token "
+          + "for this user.", null);
+    }
+    String userId = json.get("data").get("user_id").asText();
+    if (userId.isEmpty()) {
+      throw ServiceException.NOT_FOUND("User Id not found in json data. " + json.toString());
```
Comment (desouzas) [inline src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java]: Don't put the json at the end of the error here, we don't want that in the browser. Use the same error message the other implementations use.

File: src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the app access token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring the Facebook app "
+        + "access token.", null);
+    }
+
+    String appAccessToken = json.get("access_token").asText();
+    if (appAccessToken.isEmpty()) {
+      throw ServiceException.NOT_FOUND("The Facebook App token not found in "
```
Comment (desouzas) [inline src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java]: Use a similar message as the other implementations do for primary email - see Yahoo's example.

File: src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java (hunk 2 above)
Comment (desouzas) [reply src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java]: If the `user_id` is not in this response, what does that imply?

In Yahoo's case it means the yahoo app is mis-configured, is it similar here or is it not even possible?

**Comment by chinmay15** [comment]
Add description to the PR.  http://wiki.eng.zimbra.com/index.php/Code_Review_Checklist
Also format code using http://wiki.eng.zimbra.com/index.php/Zimbra_Code_Style

File: src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    String url = FacebookConstants.AUTHENTICATE_URI;
+    String queryString = "?client_id=" + this.clientId + "&client_secret="
+        + this.clientSecret + "&grant_type=client_credentials";
+    try {
+      final GetMethod request = new GetMethod(url);
+      request.setQueryString(queryString);
+      json = executeRequestForJson(request);
+    } catch (final IOException e) {
+      ZimbraLog.extensions.debug("There was an issue acquiring the app access token.", e);
+      throw ServiceException.PROXY_ERROR("There was an issue acquiring the Facebook app "
```
Comment (chinmay15) [inline src/java/com/zimbra/oauth/handlers/impl/FacebookOAuth2Handler.java]: PROXY_ERROR for IOE is not a good choice, you can use FAILURE if you don't find any suitable option.

**Pull request merged**